dulwich/repo.py

```diff
@@ -698,7 +698,12 @@ def update_shallow(self, new_shallow, new_unshallow):
             shallow.update(new_shallow)
         if new_unshallow:
             shallow.difference_update(new_unshallow)
-        self._put_named_file("shallow", b"".join([sha + b"\n" for sha in shallow]))
+        if shallow:
+            self._put_named_file(
+                "shallow", b"".join([sha + b"\n" for sha in shallow])
+            )
+        else:
+            self._del_named_file("shallow")
```

Comment:
CLI git removes this file if it is empty after a `fetch --depth=...` or `fetch --unshallow`